text/0000-export-visibility.md

```diff
@@ -0,0 +1,729 @@
+  that matches `cdylib` behavior.
+    - Initially this behavior would be gated behind a new unstable `-Z` flag.
+        - Either a new flag: `-Zhide-rust-level-exports`
+        - Or by ensuring that `-Zdefault-visibility=hidden` also applies
+          to Rust-level exports transitively linked into a `staticlib`
+    - Open question: should this behavior become the default in the future?
+    - Implementation-wise, this behavior can be seen as partial linking or
+      as object file rewriting.
+    - It seems that this behavior makes most sense for `staticlib`s, because
+      `rlib`s don't include transitive dependencies.  So when linking `rlib`s
+      (using an external, non-`rustc` linker) hiding standard library symbols
+      would still require rebuilding it with `-Zdefault-visibility=hidden`
+      as described in the "do nothing" alternative above.
+
+## Windows and `__declspec(dllexport)`
+[interposable-vs-dllexport]: #windows-and-__declspecdllexport
+
+We need to decide whether `#[export_visibility = "interposable"]` should also
+result in `__declspec((dllexport))` being added to a symbol.  See for example
+[this Stack Overflow question and answer](https://stackoverflow.com/a/25746044/24042981).
+
+Potential answers:
+
+* Don't stabilize for now (or don't support at all)
+  `#[export_visibility = "interposable"]` but still support other visibilities
+* `#[export_visibility = "interposable"]` should only control visibility
+* `#[export_visibility = "interposable"]` should control visibility
+  and also use `__declspec(dllexport)`
+
+## Interposability vs LLVM optimization passes
+[interposable-vs-llvm]: #interposability-vs-llvm-optimization-passes
+
+This RFC proposes to use `interposable` to map to
+[`SymbolVisibility::Interposable`](https://github.com/rust-lang/rust/blob/81a964c23ea4fe9ab52b4449bb166bf280035797/compiler/rustc_target/src/spec/mod.rs#L842)
+which is then mapped to
+[`llvm::Visibility::Default`](https://github.com/rust-lang/rust/blob/81a964c23ea4fe9ab52b4449bb166bf280035797/compiler/rustc_codegen_llvm/src/llvm/ffi.rs#L167).  This mimics how `interposable` is implemented and supported
+in
+[`-Zdefault-visibility=...`](https://doc.rust-lang.org/beta/unstable-book/compiler-flags/default-visibility.html).
+
+One problem here is that `llvm::Visibility::Default` is not sufficient to
+achieve actual interposability.  https://crbug.com/418073233 has one example of
+undefined behavior, but even if DSO-local global data structures were not an
+issue, then LLVM-level assumptions could still lead to undefined behavior.
+This is because the LLVM optimization passes assume that a symbol with normal
+external linkage (not weak, odr, etc) the definition it can see is the
+definition that will be actually used.  To avoid these LLVM assumptions `rustc`
+would have to enable
+[the SemanticInterposition feature](https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-fsemantic-interposition).
```

Comment:
Is there no way to enable this on a per-symbol basis? It seems like a stronger version of `#[inline(never)]` would be close to what you want.

Otherwise it sounds like we can't implement this option soundly without disabling a bunch of optimizations, so we should either remove it or mark the attribute unsafe.

Reply:
This should now be addressed by https://github.com/rust-lang/rfcs/commit/2d5eea827d64382f5b632a908e8f03d75f61dce3 which removed interposable visibility support from the text of the RFC.  (And also by https://github.com/rust-lang/rfcs/commit/3d47e56adbf98efc71ee519e73ac28ab998aaeef.)